File: cellulosesz-api/src/main/java/top/likoslupus/cellulosesz/api/text/LocalizedMessage.java
package top.likoslupus.cellulosesz.api.text;

import static top.likoslupus.cellulosesz.api.validation.TextChecks.requireNonBlank;

import static java.util.Objects.requireNonNull;

/** Platform-neutral reference to a localized message template and ordered arguments. */
public record LocalizedMessage(
        String key,
        MessageArguments arguments
) {

    public LocalizedMessage {
        key = requireNonBlank(requireNonNull(key, "key").trim(), "key");
        requireNonNull(arguments, "arguments");
    }

    public static LocalizedMessage of(String key) {
        return new LocalizedMessage(key, MessageArguments.empty());
    }

    public static LocalizedMessage of(String key, MessageArguments arguments) {
        return new LocalizedMessage(key, arguments);
    }

}

File: cellulosesz-api/src/main/kotlin/top/likoslupus/cellulosesz/api/text/LocalizedMessage.kt
package top.likoslupus.cellulosesz.api.text

import top.likoslupus.cellulosesz.api.validation.requireNonBlank

/** Platform-neutral reference to a localized message template and ordered arguments. */
@JvmRecord
public data class LocalizedMessage(
    public val key: String,
    public val arguments: MessageArguments
) {

    init {
        key.requireNonBlank { "key" }
    }

    public companion object {

        @JvmStatic
        public fun of(key: String): LocalizedMessage =
            LocalizedMessage(key, MessageArguments.empty())

        @JvmStatic
        public fun of(key: String, arguments: MessageArguments): LocalizedMessage =
            LocalizedMessage(key, arguments)

    }

}
